### ISSUE TITLE
GET_PART replication tasks hang for hours

### ISSUE THREAD

**Issue by genzgd** [ClickHouse/ClickHouse#23555]
ClickHouse 21.4.4

We have now seen this issue twice since upgrading to ClickHouse 21.4.4 about a week ago.  I don't think we saw anything quite like it previously.

On one of our clusters of 10 shards, 4 replicas, within the space of approximately 30 seconds one set GET_PART tasks will execute on multiple tables and multiple shards for relatively small parts.  This group of tasks will simply hang for several hours, until we have to restart the service to clear the replication queues which basically grows out of control (more than 100k entries waiting for the "hung" parts to complete.). 

The hung tasks have only occurred in one of our three datacenters, and since they happen on multiple nodes/shards at approximately the same time it's quite possibly related to a network glitch.

The tasks show as "currently_executing" for three or four hours.  There are no retries, exceptions, or postpone reasons.  Even on restarting the service some of these tasks remain stuck for 20 or 30 minutes, but they do seem to eventually clear and the replication_queue drains.

I've started running the log in debug mode to capture more data assuming it happens again.  There is nothing in the log referencing these stuck parts at the "information" level, our normal mode.

It seems like something happens during the fetch that leaves these tasks orphaned under certain (rare) circumstances, since we do several hundred terabytes of data per day and it's only happened twice.

**Comment by filimonov** [comment]
Please run

```
 SET allow_introspection_functions = 1;
 WITH arrayMap(x -> demangle(addressToSymbol(x)), trace) AS all SELECT thread_id, query_id, arrayStringConcat(all, '\n') AS res FROM system.stack_trace　WHERE res ILIKE '%Pool%';

 select name, value from system.settings where name like '%pool%'

 select metric, value from system.metrics where metric like 'Background%';

```

**Comment by genzgd** [comment]
I will capture these values the next time it happens (we can't run our production servers in a state where the replication queue has grown to 150k entries).  The system.stack_trace query currently returns more than 1000 entries; if there's a way to limit the query to get what you're looking for that would be helpful.

Other settings from one the dozen or so boxes where it happened (all at the same time on Friday):


```
SELECT
    name,
    value
FROM system.settings
WHERE name LIKE '%pool%'

┌─name─────────────────────────────────────────┬─value─┐
│ connection_pool_max_wait_ms                  │ 0     │
│ distributed_connections_pool_size            │ 1024  │
│ background_buffer_flush_schedule_pool_size   │ 16    │
│ background_pool_size                         │ 32    │
│ background_move_pool_size                    │ 8     │
│ background_fetches_pool_size                 │ 48    │
│ background_schedule_pool_size                │ 16    │
│ background_message_broker_schedule_pool_size │ 16    │
│ background_distributed_schedule_pool_size    │ 16    │
│ postgresql_connection_pool_size              │ 16    │
│ postgresql_connection_pool_wait_timeout      │ -1    │
└──────────────────────────────────────────────┴───────┘
```

```
SELECT
    metric,
    value
FROM system.metrics
WHERE metric LIKE 'Background%'

┌─metric──────────────────────────────────┬─value─┐
│ BackgroundPoolTask                      │    11 │
│ BackgroundFetchesPoolTask               │     0 │
│ BackgroundMovePoolTask                  │     0 │
│ BackgroundSchedulePoolTask              │     0 │
│ BackgroundBufferFlushSchedulePoolTask   │     0 │
│ BackgroundDistributedSchedulePoolTask   │     0 │
│ BackgroundMessageBrokerSchedulePoolTask │     0 │
└─────────────────────────────────────────┴───────┘
```

**Comment by genzgd** [comment]
This seems similar to #22438, but the version we are running (21.4.4) should have the backported fix.

Also the issue has not reoccurred on our clusters as of yet.

**Comment by filimonov** [comment]
> This seems similar to #22438, but the version we are running (21.4.4) should have the backported fix.
> 
> Also the issue has not reoccurred on our clusters as of yet.

Are you seeing "Memory limit exceeded" exception? Does is happen that server has all allowed memory used by clickhouse?

**Comment by genzgd** [comment]
Yes, we did have "Memory limit exceeded" exceptions at about the same time.

**Comment by genzgd** [comment]
I just checked the logs on a couple of the other affected servers and they all show "Memory limit exceeded" exceptions right about the time that the GET_PART tasks got stuck.  I'm sorry I didn't notice that before.

**Comment by genzgd** [comment]
Example: 
```
│ auto DB::StorageReplicatedMergeTree::processQueueEntry(ReplicatedMergeTreeQueue::SelectedEntryPtr)::(anonymous class)::operator()(DB::StorageReplicatedMergeTree::LogEntryPtr &) const: Code: 241, e.displayText() = DB::Exception: Memory limit (total) exceeded: would use 496.64 GiB (attempt to allocate chunk of 4331495 bytes), maximum: 452.91 GiB: (avg_value_size_hint = 49.046640715109575, avg_chars_size = 49.25596885813149, limit = 13872): (while reading column asset_id): (while reading from part /opt/data/clickhouse/store/39f/39fdf13a-ad50-4c5c-8c3f-fcb18fa83497/20210423_74929_74929_0/ from mark 12 with max_rows_to_read = 13872): While executing MergeTreeSequentialSource: Cannot fetch required block. Stream PipelineExecuting, part 0, Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x8914f0a in /usr/bin/clickhouse
1. DB::Exception::Exception<char const*, char const*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*&&, char const*&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, long&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&) @ 0x8929939 in /usr/bin/clickhouse
2. MemoryTracker::alloc(long) @ 0x89292ad in /usr/bin/clickhouse
3. MemoryTracker::alloc(long) @ 0x8929014 in /usr/bin/clickhouse
4. MemoryTracker::alloc(long) @ 0x8929014 in /usr/bin/clickhouse
5. DB::SerializationString::deserializeBinaryBulk(DB::IColumn&, DB::ReadBuffer&, unsigned long, double) const @ 0xf1f1344 in /usr/bin/clickhouse
6. DB::ISerialization::deserializeBinaryBulkWithMultipleStreams(COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, unsigned long, DB::ISerialization::DeserializeBinaryBulkSettings&, std::__1::shared_ptr<DB::ISerialization::DeserializeBinaryBulkState>&, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > >*) const @ 0xf1bfa15 in /usr/bin/clickhouse
7. DB::MergeTreeReaderWide::readData(DB::NameAndTypePair const&, COW<DB::IColumn>::immutable_ptr<DB::IColumn>&, unsigned long, bool, unsigned long, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, COW<DB::IColumn>::immutable_ptr<DB::IColumn> > > >&) @ 0xff9cc6e in /usr/bin/clickhouse
8. DB::MergeTreeReaderWide::readRows(unsigned long, bool, unsigned long, std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn> > >&) @ 0xff9bd2b in /usr/bin/clickhouse
9. DB::MergeTreeSequentialSource::generate() @ 0xffaecfa in /usr/bin/clickhouse
10. DB::ISource::tryGenerate() @ 0x1019b9e5 in /usr/bin/clickhouse
11. DB::ISource::work() @ 0x1019b5da in /usr/bin/clickhouse
12. DB::SourceWithProgress::work() @ 0x1034c8ba in /usr/bin/clickhouse
13. ? @ 0x101d4c7d in /usr/bin/clickhouse
14. DB::PipelineExecutor::executeStepImpl(unsigned long, unsigned long, std::__1::atomic<bool>*) @ 0x101d1931 in /usr/bin/clickhouse
15. DB::PipelineExecutor::executeStep(std::__1::atomic<bool>*) @ 0x101d037c in /usr/bin/clickhouse
16. DB::PullingPipelineExecutor::pull(DB::Chunk&) @ 0x101ddcaa in /usr/bin/clickhouse
17. DB::PullingPipelineExecutor::pull(DB::Block&) @ 0x101ddea0 in /usr/bin/clickhouse
18. DB::PipelineExecutingBlockInputStream::readImpl() @ 0x101cc274 in /usr/bin/clickhouse
19. DB::IBlockInputStream::read() @ 0xf0fafe5 in /usr/bin/clickhouse
20. DB::ColumnGathererStream::fetchNewBlock(DB::ColumnGathererStream::Source&, unsigned long) @ 0xf9e5f42 in /usr/bin/clickhouse
21. void DB::ColumnGathererStream::gather<DB::ColumnString>(DB::ColumnString&) @ 0xfa7cc41 in /usr/bin/clickhouse
22. DB::ColumnGathererStream::readImpl() @ 0xf9e5d53 in /usr/bin/clickhouse
23. DB::IBlockInputStream::read() @ 0xf0fafe5 in /usr/bin/clickhouse
24. DB::MergeTreeDataMergerMutator::mergePartsToTemporaryPart(DB::FutureMergedMutatedPart const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::BackgroundProcessListEntry<DB::MergeListElement, DB::MergeInfo>&, std::__1::shared_ptr<DB::RWLockImpl::LockHolderImpl>&, long, DB::Context const&, std::__1::unique_ptr<DB::IReservation, std::__1::default_delete<DB::IReservation> > const&, bool, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0xfebfe3d in /usr/bin/clickhouse
25. DB::StorageReplicatedMergeTree::tryExecuteMerge(DB::ReplicatedMergeTreeLogEntry const&) @ 0xfc64fa7 in /usr/bin/clickhouse
26. DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry&) @ 0xfc5978f in /usr/bin/clickhouse
27. ? @ 0xfcd46af in /usr/bin/clickhouse
28. DB::ReplicatedMergeTreeQueue::processEntry(std::__1::function<std::__1::shared_ptr<zkutil::ZooKeeper> ()>, std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&, std::__1::function<bool (std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>) @ 0x10025c38 in /usr/bin/clickhouse
29. DB::StorageReplicatedMergeTree::processQueueEntry(std::__1::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>) @ 0xfc884fd in /usr/bin/clickhouse
30. ? @ 0xfe00517 in /usr/bin/clickhouse
31. ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0x8957a18 in /usr/bin/clickhouse
 (version 21.4.4.30 (official build))

**Comment by genzgd** [comment]
This issue repeated, again in conjunction with "memory limit exceeded" errors

**Comment by genzgd** [comment]
We're still on 21.4.4, and the issue happened again, in connection with "memory limit exceeded" errors.

[threads.txt](https://github.com/ClickHouse/ClickHouse/files/6497421/threads.txt)

[replication_queue.txt](https://github.com/ClickHouse/ClickHouse/files/6497426/replication_queue.txt)

The replication queue is actually a TSV but Github didn't like that extension.
Note that the "stuck" GET_PART tasks in the replication_queue are now more than 2 hours old.

We had to restart this production server to clear the queues

**Comment by alexey-milovidov** [comment]
The issue has been fixed in #24483 and the fix will be available in version 21.7.
